### DISCUSSION TITLE
Using COPY FROM STDIN with list of lists as input

### DISCUSSION THREAD

**Discussion by g-plus-theworld** [psycopg/psycopg#816] [Q&A]
Hi All,

I have been trying to ingest some data to my db using COPY FROM but I am having trouble figuring out how to use it with Python lists. I just want to make sure I do not hit any bottlenecks with unnecessary loops - hence I'd rather not use copy.write_row() either.

I was able to convert the list of rows (list of lists) convert to a CSV like stream using the code below, but it seems like a lot of overhead though. In this example the list of lists is actually a list of pyodbc.Row objects, but that is on me I guess. I am only looking for a conversion between list of lists or list of tuples to ingest using COPY FROM STDIN.

```Python
    conn = psycopg.connect(conn_str)

    import csv
    import io

    with conn as c:
        cursor = c.cursor()
        data = [tuple(row) for row in rows]

        # Create a file-like object
        sio = io.StringIO()

        # Write the data to the file-like object
        writer = csv.writer(sio)
        writer.writerows(data)

        # Get the CSV data as a string
        csv_data = sio.getvalue()

        cols = get_table_col_names(table_name) # just a helper function to make the function more flexible

        copy_query = f"COPY {table_name} ({cols}) FROM STDIN WITH CSV"

        with cursor.copy(copy_query) as copy:
            copy.write(data)

        c.commit()
```

Is there a better, more concise way of transforming a list of lists to a format which then can be fed to COPY FROM STDIN? I have read the API docs and the Using Copy From and Copy To as well but I did not see any examples that would help me achieve this.

Thank you very much in advance!

**Comment by dvarrazzo** [comment]
I don't think that avoid to use write_row but using CSV instead avoid so many bottleneck. So I would start simple, with `copy.write_row()`, until it works, and then you try to measure and optimize.

That would be concise, like:

```python
    with psycopg.connect(conn_str) as c:
        cursor = c.cursor()
        with cursor.copy(copy_query) as copy:
            for row in rows:
                copy.write_row(row)
```

If you decide to use CSV and `copy.write()`, then you are just formatting a file which Postgres will have to parse, so you need to read the Postgres COPY documentation to know the right format.
